File: lab2/source/IfBranch.cpp
#include "operations.h"
#include "factory.h"
#include "interpretator.h"


class IfBranch : public Operations{
public:
    void Operation(std::stack<int> &stack, std::istream& fin) override{
        int if_counter = 1;
        std::string word;
        if(stack.top() == 0){
            while (if_counter != 0){
                if(!(fin >> word)){
                    throw std::runtime_error("incorrect cycle");
                }
                if(word == "if"){
                    if_counter += 1;
                }
                if(word == "then"){
                    fin >> word;
                    if(word == ";") if_counter -= 1;
                }
                if (word == "else"){
                    while (if_counter != 0){
                        if(!(fin >> word)){
                            throw std::runtime_error("incorrect cycle");
                        }
                        if (word != "then" && word != "else") {
                            interpretator::StackOperations(word, stack, fin);
                        }
                        if(word == "then"){
                            fin >> word;
                            if(word == ";") if_counter -= 1;
                        }
                        if(word == "else"){
                            throw std::runtime_error("incorrect cycle");
                        }
                    }
                }
            }
        } else {
            while (if_counter != 0){
                if(!(fin >> word)){
                    throw std::runtime_error("incorrect if_branch");
                }
                if(word != "then" && word != "else") {
                    interpretator::StackOperations(word, stack, fin);
                }
                if(word == "then"){
                    fin >> word;
                    if(word == ";") if_counter -= 1;
                }
                if (word == "else") {
                    while (if_counter != 0) {
                        if(!(fin >> word)){
                            throw std::runtime_error("incorrect if_branch");
                        }
                        if(word == "then"){
                            fin >> word;
                            if(word == ";") if_counter -= 1;
                        }
                        if(word == "else"){
                            throw std::runtime_error("incorrect if_branch");
                        }
                    }
                }
            }
        }
    }

};

namespace {
    Operations* createIfBranch() {
        return new IfBranch;
    }
    bool x = OperationsFactory::getInstance()->RegisterOperation(std::string("if"), createIfBranch);
}

File: jaba/lab1/src/main/java/org/example/IfBranch.java
package org.example;

import java.util.Objects;
import java.util.Scanner;
import java.util.Stack;
import java.util.NoSuchElementException;


public class IfBranch extends Operations{
    @Override
    public void Operation(Stack<Integer> stack, Scanner scanner){
        int if_counter = 1;
        String word = null;
        if(stack.peek() == 0){
            while(if_counter != 0){
                try {
                    word = scanner.next();
                } catch (NoSuchElementException e){
                    throw new NoSuchElementException("incorrect if_branch");
                }
                if(Objects.equals(word, "if")){
                    if_counter += 1;
                }
                if(Objects.equals(word, "then")){
                    try {
                        word = scanner.next();
                    } catch (NoSuchElementException e){
                        throw new NoSuchElementException("incorrect if_branch");
                    }
                    if(Objects.equals(word, ";")) if_counter -= 1;
                }
                if(Objects.equals(word, "else")){
                    while (if_counter != 0){
                        try {
                            word = scanner.next();
                        } catch (NoSuchElementException e){
                            throw new NoSuchElementException("incorrect if_branch");
                        }
                        if(!Objects.equals(word, "then") && !Objects.equals(word, "else")){
                            Interpretator.StackOperations(word, stack, scanner);
                        }
                        if(Objects.equals(word, "then")){
                            try {
                                word = scanner.next();
                            } catch (NoSuchElementException e){
                                throw new NoSuchElementException("incorrect if_branch");
                            }
                            if(Objects.equals(word, ";")) if_counter -= 1;
                        }
                        if(Objects.equals(word, "else")){
                            throw new NoSuchElementException("incorrect if_branch");
                        }
                    }
                }
            }
        } else {
            while(if_counter != 0) {
                try {
                    word = scanner.next();
                } catch (NoSuchElementException e){
                    throw new NoSuchElementException("incorrect if_branch");
                }
                if (!Objects.equals(word, "then") && !Objects.equals(word, "else")) {
                    Interpretator.StackOperations(word, stack, scanner);
                }
                if (Objects.equals(word, "then")) {
                    try {
                        word = scanner.next();
                    } catch (NoSuchElementException e){
                        throw new NoSuchElementException("incorrect if_branch");
                    }
                    if (Objects.equals(word, ";")) if_counter -= 1;
                }
                if (Objects.equals(word, "else")) {
                    while (if_counter != 0) {
                        try {
                            word = scanner.next();
                        } catch (NoSuchElementException e){
                            throw new NoSuchElementException("incorrect if_branch");
                        }
                        if (Objects.equals(word, "then")) {
                            try {
                                word = scanner.next();
                            } catch (NoSuchElementException e){
                                throw new NoSuchElementException("incorrect if_branch");
                            }
                            if (Objects.equals(word, ";")) if_counter -= 1;
                        }
                        if (Objects.equals(word, "else")) {
                            throw new NoSuchElementException("incorrect if_branch");
                        }
                    }
                }
            }
        }
    }
}
